feat(stateful-set): allow configuring `volumeClaimTemplates`

benkeil opened the pull request:
Fixes cdk8s-team/cdk8s#2183

benkeil committed:
feat: add `volumeClaimTemplates` to `kplus.StatefulSet`

Signed-off-by: Ben Keil <benkeil.pub@pm.me>

cdk8s-automation committed:
chore(deps): upgrade dev dependencies (#6385)

Upgrades project dependencies. See details in [workflow run].

[Workflow Run]: https://github.com/cdk8s-team/cdk8s-plus/actions/runs/18441834896

------

*Automatically created by projen via the "upgrade-dev-dependencies-k8s-33-main" workflow*

Signed-off-by: Ben Keil <benkeil.pub@pm.me>

benkeil committed:
fix: remove unused method

Signed-off-by: Ben Keil <benkeil.pub@pm.me>

cdk8s-automation committed:
chore(deps): upgrade dev dependencies (#6388)

Upgrades project dependencies. See details in [workflow run].

[Workflow Run]: https://github.com/cdk8s-team/cdk8s-plus/actions/runs/18460692313

------

*Automatically created by projen via the "upgrade-dev-dependencies-k8s-33-main" workflow*

Signed-off-by: Ben Keil <benkeil.pub@pm.me>

cdk8s-automation committed:
chore(deps): upgrade configuration (#6390)

Upgrades project dependencies. See details in [workflow run].

[Workflow Run]: https://github.com/cdk8s-team/cdk8s-plus/actions/runs/18470006676

------

*Automatically created by projen via the "upgrade-configuration-k8s-33-main" workflow*

Signed-off-by: Ben Keil <benkeil.pub@pm.me>

cdk8s-automation committed:
chore(deps): upgrade dev dependencies (#6394)

Upgrades project dependencies. See details in [workflow run].

[Workflow Run]: https://github.com/cdk8s-team/cdk8s-plus/actions/runs/18491339374

------

*Automatically created by projen via the "upgrade-dev-dependencies-k8s-33-main" workflow*

Signed-off-by: Ben Keil <benkeil.pub@pm.me>

cdk8s-automation committed:
chore(deps): upgrade configuration (#6399)

Upgrades project dependencies. See details in [workflow run].

[Workflow Run]: https://github.com/cdk8s-team/cdk8s-plus/actions/runs/18500936773

------

*Automatically created by projen via the "upgrade-configuration-k8s-33-main" workflow*

Signed-off-by: Ben Keil <benkeil.pub@pm.me>

cdk8s-automation committed:
chore(deps): upgrade dev dependencies (#6403)

Upgrades project dependencies. See details in [workflow run].

[Workflow Run]: https://github.com/cdk8s-team/cdk8s-plus/actions/runs/18523663856

------

*Automatically created by projen via the "upgrade-dev-dependencies-k8s-33-main" workflow*

Signed-off-by: Ben Keil <benkeil.pub@pm.me>

cdk8s-automation committed:
chore(deps): upgrade compiler dependencies (#6395)

Upgrades project dependencies. See details in [workflow run].

[Workflow Run]: https://github.com/cdk8s-team/cdk8s-plus/actions/runs/18528151867

------

*Automatically created by projen via the "upgrade-compiler-dependencies-k8s-33-main" workflow*

Signed-off-by: Ben Keil <benkeil.pub@pm.me>

cdk8s-automation committed:
chore(deps): upgrade configuration (#6405)

Upgrades project dependencies. See details in [workflow run].

[Workflow Run]: https://github.com/cdk8s-team/cdk8s-plus/actions/runs/18533358066

------

*Automatically created by projen via the "upgrade-configuration-k8s-33-main" workflow*

Signed-off-by: Ben Keil <benkeil.pub@pm.me>

cdk8s-automation committed:
chore(deps): upgrade dev dependencies (#6409)

Upgrades project dependencies. See details in [workflow run].

[Workflow Run]: https://github.com/cdk8s-team/cdk8s-plus/actions/runs/18556119076

------

*Automatically created by projen via the "upgrade-dev-dependencies-k8s-33-main" workflow*

Signed-off-by: Ben Keil <benkeil.pub@pm.me>

cdk8s-automation committed:
chore(deps): upgrade compiler dependencies (#6410)

Upgrades project dependencies. See details in [workflow run].

[Workflow Run]: https://github.com/cdk8s-team/cdk8s-plus/actions/runs/18560556728

------

*Automatically created by projen via the "upgrade-compiler-dependencies-k8s-33-main" workflow*

Signed-off-by: Ben Keil <benkeil.pub@pm.me>

iliapolo commented on docs/plus/volume.md:
```diff
@@ -36,3 +36,39 @@ const redis = pod.addContainer({
+    image: 'nginx',
+    volumeMounts: [{
+      path: '/data',
+      name: Volume.fromName(chart, "pvc-template", "data-volume"),
```
To provide a more idiomatic example, extract this to a constant:

```ts
const volume = Volume.fromName(chart, "pvc-template", "data-volume");
```

Then use `volume.name` below in the template configuration.

benkeil replied:
updated

iliapolo commented on docs/plus/volume.md:
````diff
@@ -36,3 +36,39 @@ const redis = pod.addContainer({
 redis.mount('/var/lib/redis', data);
 ```
+
+## Using PersistentVolumeClaim Templates with StatefulSets
````
Ha, we don't have specific docs for stateful sets. Note to self.

iliapolo commented on src/container.ts:
```diff
@@ -212,7 +212,7 @@ export interface SeccompProfile {
 }
 
-export interface ContainerSecutiryContextCapabilities {
+export interface ContainerSecurityContextCapabilities {
```
Unfortunately we can't do this as its a breaking change :(

benkeil replied:
reverted.

iliapolo commented on .gitignore:
```diff
@@ -78,3 +78,4 @@ docs/typescript.md
 docs/python.md
 docs/java.md
 !/.projenrc.ts
+.idea
```
You need to add this via the `.projenrc.ts` file, otherwise it gets overridden during build.

benkeil replied:
added

iliapolo commented on src/stateful-set.ts:
```diff
@@ -138,10 +152,10 @@ export class StatefulSet extends workload.Workload implements IScalable {
     this.podManagementPolicy = props.podManagementPolicy ?? PodManagementPolicy.ORDERED_READY;
     this.minReady = props.minReady ?? Duration.seconds(0);
-
+    this.volumeClaimTemplates = props.volumeClaimTemplates;
```
call `addVolumeClaimTemplate` here so we benefit from its validations.

benkeil replied:
updated

iliapolo commented on src/stateful-set.ts:
```diff
@@ -279,3 +357,16 @@ export class StatefulSetUpdateStrategy {
+   *
+   * This name should match the name of a volume mount in one of the containers.
+   */
+  readonly name: string;
```
Can we accept a `Volume` here instead? or a `VolumeMount`?

benkeil replied:
I would let it as it is. You could still use the volume as a variable and use `name: volume.name`, I would not use it like this.

I could also think about something like:

```ts
VolumeClaimTemplatesSomeSuffix.fromVolume(volume, {
  storage: Size.gibibytes(20),
  accessModes: [kplus.PersistentVolumeAccessMode.READ_WRITE_ONCE_POD],
  storageClassName: 'standard',
})
```

But to be honest my absolute favourite would be something like:

```ts
new kplus.StatefulSet(chart, 'StatefulSet', {
  containers: [
    {
      image: 'foobar',
      portNumber: 80,
      volumeMounts: [
        {
          volume: Volume.fromClaimTemplate({
            name: "data",
            storage: Size.gibibytes(20),
            accessModes: [kplus.PersistentVolumeAccessMode.READ_WRITE_ONCE_POD],
            storageClassName: 'standard',
          }),
          path: '/mnt/data',
        },
      ],
    },
  ],
})
```

iliapolo replied:
I like `Volume.fromClaimTemplate` - probably a bit more involved to implement. Don't want to block on it so i'm approving this one. Feel free to add it in a subsequent PR though as an enhancement.

iliapolo commented on src/stateful-set.ts:
```diff
@@ -279,3 +357,16 @@ export class StatefulSetUpdateStrategy {
+   * The name of the claim that the StatefulSet controller will create for each pod.
+   * This will be used to name the created PVC in the format <claim-name>-<pod-name>
+   *
+   * This name should match the name of a volume mount in one of the containers.
```
This should be validated then.

benkeil replied:
added

iliapolo commented on src/stateful-set.ts:
```diff
@@ -150,11 +164,25 @@ export class StatefulSet extends workload.Workload implements IScalable {
+    this.volumeClaimTemplates = this.volumeClaimTemplates
+      ? [...this.volumeClaimTemplates, template]
+      : [template];
+    return this;
```
This isn't consistent with our APIs. Lets return void.

benkeil replied:
maybe change for all such functions in the future. would be no breaking change.

benkeil replied:
removed

cdk8s-automation committed:
chore(deps): upgrade configuration (#6414)

Upgrades project dependencies. See details in [workflow run].

[Workflow Run]: https://github.com/cdk8s-team/cdk8s-plus/actions/runs/18565712529

------

*Automatically created by projen via the "upgrade-configuration-k8s-33-main" workflow*

Signed-off-by: Ben Keil <benkeil.pub@pm.me>

benkeil committed:
fix: remarks from PR review

Signed-off-by: Ben Keil <benkeil.pub@pm.me>

benkeil committed:
fix: better docs

Signed-off-by: Ben Keil <benkeil.pub@pm.me>

iliapolo committed:
Merge branch 'k8s-33/main' into feat/add-volume-claim-templates

iliapolo commented:
@benkeil looks good, you got some build failures though.

benkeil committed:
try to fix build errors

Signed-off-by: Ben Keil <benkeil.pub@pm.me>

iliapolo commented:
@benkeil new failures :)

benkeil committed:
fix: build error

Signed-off-by: Ben Keil <benkeil.pub@pm.me>

benkeil committed:
fix: build error

Signed-off-by: Ben Keil <benkeil.pub@pm.me>

benkeil commented:
wait until it runs in my fork

benkeil commented:
ok, build was successful

iliapolo committed:
Merge branch 'k8s-33/main' into feat/add-volume-claim-templates

cdk8s-automation commented:
## 💚 All backports created successfully

| Status | Branch | Result |
|:------:|:------:|:------|
|✅|k8s-31/main|[<img src="https://img.shields.io/github/pulls/detail/state/cdk8s-team/cdk8s-plus/6419">](https://github.com/cdk8s-team/cdk8s-plus/pull/6419)|
|✅|k8s-32/main|[<img src="https://img.shields.io/github/pulls/detail/state/cdk8s-team/cdk8s-plus/6420">](https://github.com/cdk8s-team/cdk8s-plus/pull/6420)|

### Questions ?
Please refer to the [Backport tool documentation](https://github.com/sorenlouv/backport) and see the [Github Action logs](https://github.com/cdk8s-team/cdk8s-plus/actions/runs/18594444609) for details

<!--- Backport version: 9.5.1 -->

cdk8s-automation commented:
## 💚 All backports created successfully

| Status | Branch | Result |
|:------:|:------:|:------|
|✅|k8s-31/main|[<img src="https://img.shields.io/github/pulls/detail/state/cdk8s-team/cdk8s-plus/6419">](https://github.com/cdk8s-team/cdk8s-plus/pull/6419)|
|✅|k8s-32/main|[<img src="https://img.shields.io/github/pulls/detail/state/cdk8s-team/cdk8s-plus/6420">](https://github.com/cdk8s-team/cdk8s-plus/pull/6420)|

### Questions ?
Please refer to the [Backport tool documentation](https://github.com/sorenlouv/backport) and see the [Github Action logs](https://github.com/cdk8s-team/cdk8s-plus/actions/runs/18594445661) for details

<!--- Backport version: 9.5.1 -->
